### PR TITLE
Download sources during installation and upgrade to 1.0.36

### PR DESCRIPTION
**Pull request by JimboJoe**

### CODE REVIEW

**Comment by JimboJoe** [comment]
Finally implemented dependencies download during install/upgrade with composer
Upgraded upstream version to 1.0.36

File: scripts/install [hunk 1]
```diff
@@ -42,17 +45,12 @@ ynh_app_setting_set $app mysqlpwd $dbpass
 ynh_app_setting_set $app adminusername $admin
 ynh_app_setting_set $app is_public $is_public
 
-# Download sources 
-sudo wget -q https://github.com/kanboard/kanboard/archive/v1.0.35.zip -O kanboard.zip
-
-# Uncompress
-sudo unzip -qq kanboard.zip -d ..
+# Create tmp directory and install app inside
+TMPDIR=$(ynh_mkdir_tmp)
```
Comment (mbugeia) [inline scripts/install]: This is deprecated, use mktemp since it's bash native.

File: scripts/_common.sh [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  shift 1
+
+  if [[ $USER = $(whoami) ]]; then
+    eval $@
+  else
+    # use sudo twice to be root and be allowed to use another user
+    sudo sudo -u "$USER" $@
+  fi
+}
+
```
Comment (mbugeia) [inline scripts/_common.sh]: This is so over-engineered IMO. This complicate the script for exactly no gain, the whole script is run as "admin" user by YunoHost and this is used only to run thing as admin. Plus I don't think apps script should contain such generic helper, if one is really needed it should be in yunohost core.

File: scripts/_common.sh [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+# Execute a composer command from a given directory
+# usage: composer_exec AS_USER WORKDIR COMMAND [ARG ...]
+exec_composer() {
+  local AS_USER=$1
+  local WORKDIR=$2
+  shift 2
```
Comment (mbugeia) [inline scripts/_common.sh]: shift is useless here

File: scripts/_common.sh [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  # install composer
+  curl -sS https://getcomposer.org/installer \
+    | exec_as "$AS_USER" COMPOSER_HOME="${DESTDIR}/.composer" \
+        php -- --quiet --install-dir="$DESTDIR" \
+    || ynh_die "Unable to install Composer"
+
+  # update dependencies to create composer.lock
+  exec_composer "$AS_USER" "$DESTDIR" install --no-dev \
+    || ynh_die "Unable to update application core dependencies"
+}
```
Comment (mbugeia) [inline scripts/_common.sh]: I think exec_composer and init_composer function should be merged since:
 - they are never called independently
 - they can be simplified by removing all the exec_as thing

File: scripts/_common.sh [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+# Install and initialize Composer in the given directory
+# usage: init_composer DESTDIR [AS_USER]
+init_composer() {
+  local DESTDIR=$1
+  local AS_USER=${2:-admin}
+
+  # install composer
+  curl -sS https://getcomposer.org/installer \
```
Comment (mbugeia) [inline scripts/_common.sh]: this is not the best way to install composer, the composer website provide a way more secure script here: https://getcomposer.org/doc/faqs/how-to-install-composer-programmatically.md

**Comment by JimboJoe** [comment]
Thanks for taking time to review the PR!
I took the roundcube app as a reference, which obviously seems like a bad idea... We clearly don't want to restrict hardware targets if we can avoid to do that.
You had a good idea to check the kanboard-latest.zip for dependencies... I blindly checked that URL: https://kanboard.net/kanboard-1.0.36.zip
And bingo! Dependencies are there!
I'll update the PR to revert to the first commit, updating to that link.

**Comment by JimboJoe** [comment]
Just did another commit: now downloading Kanboard sources from kanboard.net website, avoiding the need to use composer.

File: scripts/_common.sh [hunk 6]
```diff
@@ -16,53 +16,11 @@ APPLICATION_SOURCE_URL="https://github.com/kanboard/kanboard/archive/v${VERSION}
...
-    || ynh_die "Unable to install Composer"
-
-  # update dependencies to create composer.lock
-  exec_composer "$AS_USER" "$DESTDIR" install --no-dev \
-    || ynh_die "Unable to update application core dependencies"
+  archive="${DESTDIR}/application.zip"
+  wget -q -O "$archive" "$APPLICATION_SOURCE_URL" \
+    || ynh_die "Unable to download application archive"
+  # Here we process with unzip as would tar "--strip-component" option
+  unzip -qq "$archive" -d "$DESTDIR" && rm "$archive" && f=("$DESTDIR"/*) && mv "$DESTDIR"/*/* "$DESTDIR" && rm -f "${f[@]}"/.htaccess && rmdir "${f[@]}" \
```
Comment (mbugeia) [inline scripts/_common.sh]: Wow! Why is this needed? I remember having to deal with the lack of strip-component option in unzip before but I don't remember how I solved it. 
Here is how I handle the gogs zip archive for example: https://github.com/YunoHost-Apps/gogs_ynh/blob/master/scripts/_common.sh#L32-L47
If it's really needed please move it to its own function since it's not obvious how it work.

File: scripts/install [hunk 7]
```diff
@@ -46,9 +46,9 @@ ynh_app_setting_set $app adminusername $admin
 ynh_app_setting_set $app is_public $is_public
 
 # Create tmp directory and install app inside
-TMPDIR=$(ynh_mkdir_tmp)
+TMPDIR=`mktemp -d`
```
Comment (mbugeia) [inline scripts/install]: I'm a little picky here but could you use $(command) instead of backquotes theses are deprecated. Also consistency with the rest of the script.

File: scripts/_common.sh (hunk 6 above)
Comment (JimboJoe) [inline scripts/_common.sh]: When looking for a substitute to --strip-component, I also stumbled upon rsync-based solutions. I do prefer using bash-only when possible.
But I agree it's ugly, so I'll make it cleaner in a specific function.

File: scripts/install (hunk 7 above)
Comment (JimboJoe) [inline scripts/install]: OK, I fell into old habits here :-)

File: scripts/_common.sh (hunk 6 above)
Comment (mbugeia) [inline scripts/_common.sh]: Oh ok so it was probably Gogs. But since rsync is installed in debian like unzip why bother? I prefer not to have complicated `rm -f` while rsync do the job very well.

**Comment by mbugeia** [comment]
Ok for me, tested the upgrade on my production instance yesterday without issue. I'd like to have another person to test it before merging since it's an official app.
